btcec/signature.go

```diff
@@ -396,3 +401,134 @@ func RecoverCompact(curve *KoblitzCurve, signature,
 
 	return key, ((signature[0] - 27) & 4) == 4, nil
 }
+
+var one = big.NewInt(1)
+
+// SignRFC6979 generates a deterministic ECDSA signature according to RFC 6979 and BIP 62.
+func SignRFC6979(privateKey *PrivateKey, hash []byte) (*Signature, error) {
+
+	privkey := privateKey.ToECDSA()
+	N := order
+	k := NonceRFC6979(privkey.D.Bytes(), hash)
+	inv := new(big.Int).ModInverse(k, N)
+	r, _ := privkey.Curve.ScalarBaseMult(k.Bytes())
+	r.Mod(r, N)
+
+	if r.Sign() == 0 {
+		return nil, errors.New("calculated R is zero")
+	}
+
+	e := hashToInt(hash, privkey.Curve)
+	s := new(big.Int).Mul(privkey.D, r)
+	s.Add(s, e)
+	s.Mul(s, inv)
+	s.Mod(s, N)
+
+	if s.Cmp(halforder) == 1 {
+		s = new(big.Int).Sub(order, s)
+	}
+
+	return &Signature{R: r, S: s}, nil
+}
+
+// NonceRFC6979 generates a ECDSA nonce (`k`) deterministically according to RFC 6979.
+// It takes a 32-byte hash as an input and returns 32-byte nonce to be used in ECDSA algorithm.
+func NonceRFC6979(privkey []byte, hash []byte) *big.Int {
+
+	q := order
+	x := new(big.Int).SetBytes(privkey)
+	alg := fastsha256.New
+
+	qlen := q.BitLen()
+	holen := alg().Size()
+	rolen := (qlen + 7) >> 3
+	bx := append(int2octets(x, rolen), bits2octets(hash, q, qlen, rolen)...)
+
+	// Step B
+	v := bytes.Repeat([]byte{0x01}, holen)
```

Comment:
I'd suggest using a package-level variable like:

```Go
// oneInitializer is used to fill a byte slice with byte 0x01.  It is provided
// here to avoid the need to create it multiple times.
var oneInitializer = []byte{0x01}
```

Then using it here:

```Go
// Step B.
v := bytes.Repeat(oneInitializer, holen)
```

The reason is that the current code will create a new byte slice for what is essentially a constant value on every call to the function.  In general, generating extra garbage can quickly lead to really slowing Go programs down, so I like to make an effort to avoid extra allocations wherever possible.